File: programming/lab2/src/main/java/moves/special/ShadowBall.java
package moves.special;

import ru.ifmo.se.pokemon.*;

/**
 * Shadow Ball deals damage and has a 20% chance of lowering the target's Special Defense by one stage.
 */
public class ShadowBall extends SpecialMove {
    /**
     * Шанс на понижение SPECIAL_DEFENSE.
     * Необходим для тестов
     */
    private double specialDefenseLowerChance = 0.2;

    public ShadowBall() {
        super(Type.GHOST, 80, 100);
    }

    public ShadowBall(double acc) {
        this();
        specialDefenseLowerChance = acc;
    }

    @Override
    protected boolean checkAccuracy(Pokemon pokemon, Pokemon pokemon1) {
        return true;
    }

    @Override
    protected String describe() {
        return "использует Shadow Ball";
    }

    @Override
    protected void applyOppEffects(Pokemon pokemon) {
        System.out.println(specialDefenseLowerChance);
        pokemon.addEffect(new Effect().chance(specialDefenseLowerChance).stat(Stat.SPECIAL_DEFENSE, -1));
    }
}

File: programming/lab2/app/src/main/kotlin/org/itmo/moves/special/ShadowBall.kt
package org.itmo.moves.special

import ru.ifmo.se.pokemon.*

class ShadowBall : SpecialMove(Type.GHOST, 80.0, 100.0) {
    override fun applyOppEffects(p0: Pokemon) {
        p0.addEffect(Effect().chance(0.2).stat(Stat.SPECIAL_DEFENSE, -1).turns(-1))
    }

    override fun checkAccuracy(p0: Pokemon?, p1: Pokemon?): Boolean {
        return true
    }

    override fun describe(): String {
        return "использует Shadow Ball"
    }
}